neco-exporter/pkg/exporter/exporter.go

```diff
@@ -0,0 +1,145 @@
+package exporter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/VictoriaMetrics/metrics"
+)
+
+type Exporter struct {
+	scope      string
+	port       int
+	collectors []Collector
+	interval   time.Duration
+}
+
+func NewExporter(scope string, port int, collectors []Collector, interval time.Duration) *Exporter {
+	return &Exporter{
+		scope:      scope,
+		port:       port,
+		collectors: collectors,
+		interval:   interval,
+	}
+}
+
+func (e *Exporter) Start(ctx context.Context) error {
+	runErr := make(chan error)
+	go func() {
+		if err := e.Run(ctx); err != nil {
+			slog.ErrorContext(ctx, "failed to scrape metrics", slog.Any("error", err))
+			runErr <- err
+		}
+	}()
+
+	addr := fmt.Sprintf("0.0.0.0:%d", e.port)
+	slog.InfoContext(ctx, "start metrics server", slog.Any("address", addr))
+
+	server := http.Server{
+		Addr:              addr,
+		ReadHeaderTimeout: 5 * time.Second,
+		ReadTimeout:       5 * time.Second,
+		WriteTimeout:      60 * time.Second,
+		IdleTimeout:       120 * time.Second,
+	}
+
+	http.HandleFunc("/metrics", func(w http.ResponseWriter, req *http.Request) {
+		metrics.WritePrometheus(w, false)
+	})
+
+	serveErr := make(chan error)
+	go func() {
+		if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			slog.ErrorContext(ctx, "metrics server stopped", slog.Any("error", err))
+			serveErr <- err
+		}
+	}()
+
+	select {
+	case err := <-runErr:
+		return err
```

Comment:
I think the server should be stopped by `Shutdown()` before returning.
And regarding the error, I think it's sufficient to either log it or return it, but not both.